sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1192,11 +1195,23 @@ class Analyzer(
    * @see https://issues.apache.org/jira/browse/SPARK-19737
    */
   object LookupFunctions extends Rule[LogicalPlan] {
-    override def apply(plan: LogicalPlan): LogicalPlan = plan.transformAllExpressions {
-      case f: UnresolvedFunction if !catalog.functionExists(f.name) =>
-        withPosition(f) {
-          throw new NoSuchFunctionException(f.name.database.getOrElse("default"), f.name.funcName)
-        }
+    override def apply(plan: LogicalPlan): LogicalPlan = {
+      val catalogFunctionNameSet = new mutable.HashSet[FunctionIdentifier]()
+      plan.transformAllExpressions {
+        case f: UnresolvedFunction if catalogFunctionNameSet.contains(f.name) => f
+        case f: UnresolvedFunction if catalog.functionExists(f.name) =>
+          catalogFunctionNameSet.add(normalizeFuncName(f.name))
+          f
+        case f: UnresolvedFunction =>
+          withPosition(f) {
+            throw new NoSuchFunctionException(f.name.database.getOrElse("default"),
+              f.name.funcName)
+          }
+      }
+    }
+
+    private def normalizeFuncName(name: FunctionIdentifier): FunctionIdentifier = {
+      FunctionIdentifier(name.funcName.toLowerCase(Locale.ROOT), name.database)
```

Comment:
`name.database.getOrElse("default")`?

Reply:
the FunctionIdentifier's signature for database is Option, it is not string. since we are just used in this local cache, I think it is ok to not convert to "default" string. I saw when we do the  registerFunction in FunctionRegistry.scala, we didn't put the "default" in normalizeFuncName either. What do you think? thanks.

Reply:
@viirya Hello Simon, I will leave this as it is for now, but if you think it is better to have the Option(name.database.getOrElse("default")) in the catalogFunctionNameSet, let me know. Thanks.

Reply:
Ah, sorry for replying late. What I thought is if two `FunctionIdentifier`s, one is with default database name `Some("default")`, another is `None`. They should be equal to each other here.

I actually mean `name.database.orElse(Some("default"))`.

Reply:
@viirya Shouldn't we be using the current database if database is not specified ? I am trying to understand why we should use "default" here ?

Reply:
@dilipbiswal Ah, right. Sorry it should be normalized as `name.database.orElse(Some(catalog.getCurrentDatabase))`.

Reply:
@viirya Thank you.

Reply:
@kevinyu98 For built-in functions, we don't need to normalize their database name.

Rethink about it, actually here it is not for function resolution, I think it is OK to leave it as `name.database`.

Reply:
@viirya @kevinyu98 We need to check what happens in the following case .
```
use currentdb;
select currentdb.function1(), function1() from ....
```
In this case, the 2nd function should be resolved from the local cache if this optimization
were to work. If we just use name.database  instead of defaulting to current database , will it still happen ?

Reply:
@dilipbiswal @viirya Thanks for pointing this out. If we just use the name.database, the cache will store "None" for the database name, the 2nd function will not resolved from the local cache. We need to use the catalog.getCurrentDatabase for the database name in the cache.
After running more test cases, I think it is better to cache the external function name only, not include the build-in function. If we all agree this approach, I can submit the code for review.

Reply:
For built-in functions, it may no a big deal if we don't find it in this cache. It should be very fast to query built-in functions. I remember the main issue of this ticket is external function lookup where it means more loading on connection with metastore.

Reply:
i agree @viirya